packages/cli/src/ActivationErrors.service.ts

```diff
@@ -0,0 +1,52 @@
+import { Service } from 'typedi';
+import { CacheService } from './services/cache.service';
+import { jsonParse } from 'n8n-workflow';
+
+type ActivationErrors = {
+	[workflowId: string]: string; // error message
+};
+
+@Service()
+export class ActivationErrorsService {
+	private readonly cacheKey = 'workflow-activation-errors';
+
+	constructor(private readonly cacheService: CacheService) {}
+
+	async set(workflowId: string, errorMessage: string) {
+		const errors = await this.getAll();
+
+		errors[workflowId] = errorMessage;
+
+		await this.cacheService.set(this.cacheKey, JSON.stringify(errors));
```

Comment:
Thanks for addressing this to use a single key - but I think we might have a bigger problem now.

See, on line 16, we perform `getAll`. When we talk about workflow activation, we nearly try to activate all workflows at the same time, right? So if 2 workflows fail, we could end up in a scenario like this:

- Error for workflow A performs `getAll` - it returns an empty list
- Error for workflow B also performs `getAll` - it also gets an empty list. The two calls are cocnurrent.
- Error from workflow A is set, like `{"workflowA": "Trigger failed"}`
- Error from workflow B is set, like `{"workflowB": "Trigger failed"}`
- Error from workflow A is saved; Redis contains `{"workflowA": "Trigger failed"}`
- Error from workflow B is then saved, overriding what A just did, and now Redis contains only `{"workflowB": "Trigger failed"}`

To address this I think it's best to improve the `cacheService` to allow for atomic hash operations using [Redis Hashes](https://redis.io/docs/data-types/hashes/). WDYT?

Reply:
Interesting!

> When we talk about workflow activation, we nearly try to activate all workflows at the same time, right?

So activations are sequential: in `ActiveWorkflowRunner.addActiveWorkflows()` we loop over them and on each iteration we call `ActiveWorkflowRunner.add()` which sets the error via `ActiveErrorsService.set()` which internally calls `ActiveErrorsService.getAll()`. If they are sequential, your scenario might not be an issue during activations?

> improve the cacheService to allow for atomic hash operations using [Redis Hashes](https://redis.io/docs/data-types/hashes/)

At first glance the caching packages we're using don't seem to support Redis hashes? I might be wrong - I'll take a look.

Reply:
You seem to be right, the lib we picked does not seem to support this. 

In terms of sequential activation, this seems right, so not as bad, but still possible. It'd be great if we could prevent this loophole but it doesn't seem possible with the current approach.

Reply:
Given this would require switching libs, we agreed this could be addressed in future but it should be fine for now.